File: EmuSen/Cores/Nintendo/Mars - N64/Memory/Joybus.cs
using System;

namespace EmuSen.Cores.Nintendo.Mars.Memory
{
    // The PIF walking the command block a game left in PIF RAM, channel by channel - see Mars_Serial.md §3.
    public static class Joybus
    {
        public const byte End = 0xFE;
        public const byte Skip = 0xFF;
        public const byte Pad = 0xFD;

        // A length of zero is a channel with nothing to ask, which is not the same as the info command of the same number - see §3.
        public const byte Silent = 0x00;

        // Set in the length byte when nothing answered, and beside it when the reply was longer than the room for it - see §3.3.
        public const byte NoReply = 0x80;
        public const byte OverRun = 0x40;

        private const byte Info = 0x00, Reset = 0xFF, State = 0x01, PakRead = 0x02, PakWrite = 0x03;

        // The block runs to its end byte or to the last byte of PIF RAM, whichever comes first - see §3.
        public static void Run(byte[] ram, Controller[] ports, SaveChip? cartridge = null)
        {
            int at = 0;
            int channel = 0;

            while (at < ram.Length)
            {
                byte length = ram[at];

                if (length == End) break;

                // A zero says this channel has no command; padding says nothing at all - see §3.
                if (length == Silent)
                {
                    channel++;
                    at++;
                    continue;
                }

                if (length is Pad or Skip)
                {
                    at++;
                    continue;
                }

                if (++at >= ram.Length || ram[at] == End) break;

                int send = length & 0x3F;
                int receive = ram[at] & 0x3F;
                int lengthAt = at;
                int command = ++at;

                if (command + send > ram.Length || command + send + receive > ram.Length) break;

                Answer(ram, ports, cartridge, channel, command, send, receive, lengthAt);

                at = command + send + receive;
                channel++;
            }

            // The PIF clears the byte it was started by, so a block runs once - see §2.
            ram[^1] = 0;
        }

        private static void Answer(byte[] ram, Controller[] ports, SaveChip? cartridge, int channel, int command, int send, int receive, int lengthAt)
        {
            Controller? port = channel < ports.Length ? ports[channel] : null;
            byte id = send > 0 ? ram[command] : Skip;
            int wrote = 0;

            // Every channel past the four ports reaches the cartridge, as the FPGA core routes them - see Mars_Save.md §2.
            bool answered = channel >= ports.Length
                ? cartridge != null && cartridge.AnswerJoybus(ram, command, send, receive, out wrote)
                : port is { Present: true } && Answered(ram, port, id, command, send, receive, out wrote);

            if (!answered)
            {
                ram[lengthAt] = (byte)(NoReply | receive);
                return;
            }

            ram[lengthAt] = (byte)((wrote > receive ? OverRun : 0) | receive);
        }

        // The controller's two commands, and the pak's two when one is in the slot - see §3.2 and Mars_Save.md §5.
        private static bool Answered(byte[] ram, Controller port, byte id, int command, int send, int receive, out int wrote)
        {
            wrote = 0;

            switch (id)
            {
                case Info:
                case Reset:
                    wrote = 3;
                    Reply(ram, command + send, receive, 0x05, 0x00, (byte)(port.Pak != null ? 0x01 : 0x02));
                    return true;

                case State:
                    wrote = 4;
                    Reply(ram, command + send, receive,
                        (byte)(port.Buttons >> 8), (byte)port.Buttons, (byte)port.StickX, (byte)port.StickY);
                    return true;

                case PakRead:
                    if (port.Pak is null || send < 3) return false;
                    wrote = ControllerPak.ChunkSize + 1;
                    Reply(ram, command + send, receive, PakChunk(port.Pak, ram, command));
                    return true;

                case PakWrite:
                    if (port.Pak is null || send < 3 + ControllerPak.ChunkSize) return false;
                    var data = ram.AsSpan(command + 3, ControllerPak.ChunkSize);
                    port.Pak.Write(ControllerPak.Address(ram[command + 1], ram[command + 2]), data);
                    wrote = 1;
                    Reply(ram, command + send, receive, ControllerPak.DataCrc(data));
                    return true;

                default:
                    return false;
            }
        }

        // Thirty-two bytes and their CRC.
        private static byte[] PakChunk(ControllerPak pak, byte[] ram, int command)
        {
            var reply = new byte[ControllerPak.ChunkSize + 1];
            pak.Read(ControllerPak.Address(ram[command + 1], ram[command + 2]), reply.AsSpan(0, ControllerPak.ChunkSize));
            reply[^1] = ControllerPak.DataCrc(reply.AsSpan(0, ControllerPak.ChunkSize));
            return reply;
        }

        // A reply longer than the room for it is cut; a shorter one leaves the rest of the room alone - see §3.3.
        public static void Reply(byte[] ram, int at, int receive, params byte[] bytes)
        {
            for (int i = 0; i < bytes.Length && i < receive; i++) ram[at + i] = bytes[i];
        }
    }
}

File: EmuSen/Cores/Nintendo/MarsRT - N64/src/memory/joybus.rs
//! The PIF walking the command block a game left in PIF RAM, channel by channel: the C# `Joybus`. See Mars_Serial.md §3.

use crate::memory::controller::{CHUNK_SIZE, Controller, ControllerPak};
use crate::memory::save::SaveChip;

pub const END: u8 = 0xFE;
pub const SKIP: u8 = 0xFF;
pub const PAD: u8 = 0xFD;
pub const SILENT: u8 = 0x00;
pub const NO_REPLY: u8 = 0x80;
pub const OVER_RUN: u8 = 0x40;

const INFO: u8 = 0x00;
const RESET: u8 = 0xFF;
const STATE: u8 = 0x01;
const PAK_READ: u8 = 0x02;
const PAK_WRITE: u8 = 0x03;

/// `Run`: to the end byte or the last byte of PIF RAM, whichever comes first.
pub fn run(ram: &mut [u8; 64], ports: &mut [Controller; 4], cartridge: &mut SaveChip) {
    let mut at = 0usize;
    let mut channel = 0usize;
    while at < ram.len() {
        let length = ram[at];
        if length == END {
            break;
        }
        if length == SILENT {
            channel += 1;
            at += 1;
            continue;
        }
        if length == PAD || length == SKIP {
            at += 1;
            continue;
        }
        at += 1;
        if at >= ram.len() || ram[at] == END {
            break;
        }
        let send = (length & 0x3F) as usize;
        let receive = (ram[at] & 0x3F) as usize;
        let length_at = at;
        at += 1;
        let command = at;
        if command + send > ram.len() || command + send + receive > ram.len() {
            break;
        }
        answer(ram, ports, cartridge, channel, command, send, receive, length_at);
        at = command + send + receive;
        channel += 1;
    }
}

#[allow(clippy::too_many_arguments)]
fn answer(ram: &mut [u8; 64], ports: &mut [Controller; 4], cartridge: &mut SaveChip, channel: usize, command: usize, send: usize, receive: usize, length_at: usize) {
    let id = if send > 0 { ram[command] } else { SKIP };
    let mut wrote = 0usize;
    let answered = if channel >= ports.len() {
        cartridge.answer_joybus(ram, command, send, receive, &mut wrote)
    } else {
        let port = &mut ports[channel];
        port.present && answered(ram, port, id, command, send, receive, &mut wrote)
    };
    if !answered {
        ram[length_at] = NO_REPLY | receive as u8;
        return;
    }
    ram[length_at] = (if wrote > receive { OVER_RUN } else { 0 }) | receive as u8;
}

fn answered(ram: &mut [u8; 64], port: &mut Controller, id: u8, command: usize, send: usize, receive: usize, wrote: &mut usize) -> bool {
    *wrote = 0;
    match id {
        INFO | RESET => {
            *wrote = 3;
            let pak = if port.pak.is_some() { 0x01 } else { 0x02 };
            reply(ram, command + send, receive, &[0x05, 0x00, pak]);
            true
        }
        STATE => {
            *wrote = 4;
            reply(ram, command + send, receive, &[(port.buttons >> 8) as u8, port.buttons as u8, port.stick_x as u8, port.stick_y as u8]);
            true
        }
        PAK_READ => {
            let Some(pak) = port.pak.as_ref() else { return false };
            if send < 3 {
                return false;
            }
            *wrote = CHUNK_SIZE + 1;
            let mut chunk = [0u8; CHUNK_SIZE + 1];
            pak.read(ControllerPak::address(ram[command + 1], ram[command + 2]), &mut chunk[..CHUNK_SIZE]);
            chunk[CHUNK_SIZE] = ControllerPak::data_crc(&chunk[..CHUNK_SIZE]);
            reply(ram, command + send, receive, &chunk);
            true
        }
        PAK_WRITE => {
            let Some(pak) = port.pak.as_mut() else { return false };
            if send < 3 + CHUNK_SIZE {
                return false;
            }
            let mut data = [0u8; CHUNK_SIZE];
            data.copy_from_slice(&ram[command + 3..command + 3 + CHUNK_SIZE]);
            pak.write(ControllerPak::address(ram[command + 1], ram[command + 2]), &data);
            *wrote = 1;
            reply(ram, command + send, receive, &[ControllerPak::data_crc(&data)]);
            true
        }
        _ => false,
    }
}

/// `Reply`: a reply longer than the room for it is cut; a shorter one leaves the rest alone.
pub fn reply(ram: &mut [u8; 64], at: usize, receive: usize, bytes: &[u8]) {
    for (i, &b) in bytes.iter().enumerate().take(receive) {
        ram[at + i] = b;
    }
}
